src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -105,62 +108,180 @@ public void run() {
+              }
+            },
+            BROWSE_SERVICE_TIMEOUT,
+            TimeUnit.MILLISECONDS);
+
+    mainThreadHandler.postDelayed(timeoutRunnable, RESOLVE_SERVICE_TIMEOUT);
+  }
+
+  private static class NsdServiceFinderAndResolver implements NsdManager.DiscoveryListener {
+    private static final String TAG = NsdServiceFinderAndResolver.class.getSimpleName();
+
+    private final NsdManager nsdManager;
+    private final NsdServiceInfo targetServiceInfo;
+    private final long callbackHandle;
+    private final long contextHandle;
+    private final ChipMdnsCallback chipMdnsCallback;
+    private final Runnable timeoutRunnable;
+    private final MulticastLock multicastLock;
+    private final Handler mainThreadHandler;
+    @Nullable private final NsdManagerResolverAvailState nsdManagerResolverAvailState;
+
+    public NsdServiceFinderAndResolver(
+        final NsdManager nsdManager,
+        final NsdServiceInfo targetServiceInfo,
+        final long callbackHandle,
+        final long contextHandle,
+        final ChipMdnsCallback chipMdnsCallback,
+        final Runnable timeoutRunnable,
+        final MulticastLock multicastLock,
+        final Handler mainThreadHandler,
+        final NsdManagerResolverAvailState nsdManagerResolverAvailState) {
+      this.nsdManager = nsdManager;
+      this.targetServiceInfo = targetServiceInfo;
+      this.callbackHandle = callbackHandle;
+      this.contextHandle = contextHandle;
+      this.chipMdnsCallback = chipMdnsCallback;
+      this.timeoutRunnable = timeoutRunnable;
+      this.multicastLock = multicastLock;
+      this.mainThreadHandler = mainThreadHandler;
+      this.nsdManagerResolverAvailState = nsdManagerResolverAvailState;
     }
 
-    this.nsdManager.resolveService(
-        serviceInfo,
-        new NsdManager.ResolveListener() {
-          @Override
-          public void onResolveFailed(NsdServiceInfo serviceInfo, int errorCode) {
-            Log.w(
-                TAG,
-                "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
-            chipMdnsCallback.handleServiceResolve(
-                instanceName, serviceType, null, null, 0, null, callbackHandle, contextHandle);
+    public void start() {
+      this.nsdManager.discoverServices(
+          targetServiceInfo.getServiceType(), NsdManager.PROTOCOL_DNS_SD, this);
+    }
 
-            if (multicastLock.isHeld()) {
-              multicastLock.release();
+    @Override
+    public void onServiceFound(NsdServiceInfo service) {
+      Log.d(TAG, "onServiceFound: " + service);
+      if (targetServiceInfo.getServiceName().equals(service.getServiceName())) {
+        Log.d(TAG, "onServiceFound: found target service " + service);
 
-              if (nsdManagerResolverAvailState != null) {
-                nsdManagerResolverAvailState.signalFree();
+        if (nsdManagerResolverAvailState != null) {
+          nsdManagerResolverAvailState.acquireResolver();
```

Comment:
can this call ever fail?

Reply:
I think it is possible that a call to condition.await() inside acquireResolver() throws an InterruptedException if the thread is interrupted. 

But acquireResolver does not throw or return that error at the moment and I'd like to keep the focus on the new change I am introducing here. So, I'll modify acquireResolver in a separate PR later.